File: packages/cli/src/ckpt/commands/commit.py
"""ckpt commit — commit with reasoning capture."""

from __future__ import annotations

import subprocess
import sys

import click

from ckpt.core import git
from ckpt.core.reasoning import resolve_reasoning
from ckpt.storage.local import LocalStore


@click.command(
    context_settings={"ignore_unknown_options": True, "allow_extra_args": True},
)
@click.option("-m", "--message", default=None, help="Commit message.")
@click.option("--reasoning", default=None, help="Why this change was made.")
@click.pass_context
def commit(ctx: click.Context, message: str | None, reasoning: str | None):
    """Commit staged changes and record reasoning."""
    store = LocalStore()

    # Resolve reasoning from flag → stdin → staging
    final_reasoning = resolve_reasoning(flag_value=reasoning, store=store)

    # Build git commit args
    git_args = ["commit"]
    if message:
        git_args.extend(["-m", message])
    # Forward any extra args (e.g. --amend, --no-edit, etc.)
    git_args.extend(ctx.args)

    try:
        git.run(*git_args)
    except subprocess.CalledProcessError as e:
        sys.exit(e.returncode)

    # Capture commit info
    commit_hash = git.get_head_hash()
    parent_hash = git.get_parent_hash()
    author = git.get_author()
    files = git.get_committed_files(commit_hash)

    # Save reasoning record
    store.add_reasoning(
        commit_hash=commit_hash,
        reasoning=final_reasoning,
        author=author,
        files=files,
        parent_hash=parent_hash,
    )
    store.clear_staging()
    store.close()

    if final_reasoning:
        click.echo(f"Reasoning recorded for {commit_hash[:8]}: {final_reasoning}")
    else:
        click.echo(f"Commit {commit_hash[:8]} recorded (no reasoning provided)")

File: packages/cli/src/commands/commit.ts
import { Command } from 'commander';
import { run, getHeadHash, getParentHash, getAuthor, getCommittedFiles, getRepoUrl } from '../core/git.js';
import { requireInit } from '../core/config.js';
import { resolveReasoning } from '../core/reasoning.js';
import { detectAgent } from '../core/agent.js';
import { LocalStore } from '../storage/local.js';

export const commit = new Command('commit')
  .description('Commit staged changes with reasoning')
  .option('-m, --message <msg>', 'commit message')
  .option('--reasoning <text>', 'reasoning for this commit')
  .allowUnknownOption(true)
  .action((opts: { message?: string; reasoning?: string }, cmd: Command) => {
    requireInit();
    const store = new LocalStore();
    const reasoning = resolveReasoning(opts.reasoning, store);

    const gitArgs = ['commit'];
    if (opts.message) gitArgs.push('-m', opts.message);
    // Pass through any extra args
    const extra = cmd.args ?? [];
    gitArgs.push(...extra);

    const result = run(gitArgs);
    if (result.exitCode !== 0) {
      store.close();
      process.exit(result.exitCode);
    }

    const commitHash = getHeadHash();
    const parentHash = getParentHash();
    const author = getAuthor();
    const files = getCommittedFiles(commitHash);
    const repoUrl = getRepoUrl();
    const agent = detectAgent();

    store.addReasoning(commitHash, reasoning, author, files, parentHash, { agent }, repoUrl);
    store.clearStaging();
    store.close();

    console.log(`Committed ${commitHash.slice(0, 7)}`);
    if (reasoning) console.log(`  reasoning: ${reasoning}`);
  });
